rails/README.md

```diff
@@ -0,0 +1,208 @@
+- Use `link_to` for GET requests, and `button_to` for other HTTP verbs.
+- Use new-style `validates :name, presence: true` validations, and put all validations for a given column together.
+  [Example][validations].
+- [Add foreign key constraints][fkey] in migrations.
+- Avoid bypassing validations with methods like `save(validate: false)`, `update_attribute`, and `toggle`.
+- Avoid instantiating more than one object in controllers.
+- Avoid naming methods after database columns in the same class.
+- Don't change a migration after it has been merged into master if the desired change can be solved with another
+  migration.
+- Don't reference a model class directly from a view.
+- Don't return false from `ActiveModel` callbacks, but instead raise an exception.
+- Don't use instance variables in partials. Pass local variables to partials from view templates.
+- Don't use SQL or SQL fragments (`where('inviter_id IS NOT NULL')`) outside of models.
+- Generate necessary [Spring binstubs] for the project, such as `rake` and `rspec`, and add them to version control.
+- If there are default values, set them in migrations.
+- Keep `db/schema.rb` or `db/development_structure.sql` under version control.
+- Use only one instance variable in each view.
+- Use SQL, not `ActiveRecord` models, in migrations.
+- Use the [`.ruby-version`] file convention to specify the Ruby version and patch level for a project.
+- Use `_url` suffixes for named routes in mailer views and [redirects]. Use `_path` suffixes for named routes everywhere
+  else.
+- Validate the associated `belongs_to` object (`user`), not the database column (`user_id`).
+- Use `db/seeds.rb` for data that is required in all environments.
+- Use `dev:prime` rake task for development environment seed data.
+- Prefer `cookies.signed` over `cookies` to [prevent tampering].
+- Prefer `Time.current` over `Time.now`
+- Prefer `Date.current` over `Date.today`
+- Prefer `Time.zone.parse("2014-07-04 16:05:37")` over `Time.parse("2014-07-04 16:05:37")`
+- Use `ENV.fetch` for environment variables instead of `ENV[]`so that unset environment variables are detected on
+  deploy.
+- [Use blocks][date-block] when declaring date and time attributes in FactoryBot factories.
+- Use `touch: true` when declaring `belongs_to` relationships.
+
+[date-block]: /ruby/sample_2.rb#L10
+[fkey]: http://robots.thoughtbot.com/referential-integrity-with-foreign-keys
+[`.ruby-version`]: https://gist.github.com/fnichol/1912050
+[redirects]: http://www.w3.org/Protocols/rfc2616/rfc2616-sec14.html#sec14.30
+[spring binstubs]: https://github.com/sstephenson/rbenv/wiki/Understanding-binstubs
+[prevent tampering]: http://blog.bigbinary.com/2013/03/19/cookies-on-rails.html
+[class constant in association]: /rails/sample.rb
+[order-associations]: /rails/sample.rb#L2-L4
+[validations]: /rails/sample.rb#L6
+[`app/views/application`]: http://railscasts.com/episodes/269-template-inheritance
+
+## Starting a new Rails app
```

Comment:
I wonder if these sub headers should be split up into their own files. Maybe not though since people are probably less likely to read the other files and more likely to read things in the same file? What do you think?

Reply:
I think a good delineation could be guidelines vs. how-tos.

I feel like guidelines, even if they're split up by section, should live in the main `README.md` file for the guide.

But how-tos make sense as separate documents, with a `How To...` section in the main `README.md` linking to them.

For this guide, that might look like:

```md

## How To...

 - [Start a new Rails app](./starting_a_new_rails_app.md)
 - [Feature-test JavaScript in a Rails app](./feature_testing_rails_app_javascript.md)

```

Thoughts on that structure?

Reply:
I dig it!

Reply:
Cool 🙂 I split out the `ruby` and `rails` how-tos into sub-files and I also added a `_template` guide we can use as a reference of the expected structure (37b3b7f742ecbf49a06f97f2313c9efbff3c7e69)